Improve Nginx configuration

VirtuBox opened the pull request:
* Add php upstream example 
* use standard $args instead of $query_strings (this is the same)
* improve static files cache directives
* harden security

franzliedke commented:
@VirtuBox Thank you for the pull request!

Could you please provide more information about the *why* behind each of these changes?
In general, it is also much easier to handle / discuss these things when you separate unrelated changes, optimally in several PRs. For example, the whitespace changes...

VirtuBox commented:
> @VirtuBox Thank you for the pull request!
> 
> Could you please provide more information about the _why_ behind each of these changes?
> In general, it is also much easier to handle / discuss these things when you separate unrelated changes, optimally in several PRs. For example, the whitespace changes...

Hello @franzliedke,
I apologize about my PR style and I totally agree about the fact it would be better to split changes in several commits or PRs.

To provide more information about the changes I have done : 
- Lines 1-27 : I added a php upstream example, the location block to process php files. and the minimal nginx vhost to use with this configuration file.  This example is based on the nginx.conf.sample used by Magento 2.X. 

- Line 18 : I replaced `$query_strings`  by `$args` because the second one is used on the most part of nginx configuration examples (WordPress, Joomla, etc..). I had myself to check  the difference between them on nginx.org  

- Line 37-52 : I replaced `add_header Cache-Control` by the directive `expires` because it's the Nginx directive used to set the `Cache-Control` header. 
I also added the directives `access_log off;` and `log_not_found off;` to disable access and error logs related to static files, it reduce a lot the amount of logs created by Nginx.
I split the original block in two blocks to set a shorter expiration on css and js assets.

- Line 53-56 : I added a block to deny access to hidden files and directories. Even if flarum use the public directory as webroot, it's always a good practice to deny access on hidden files/directory like .git, .env. .composer, .htaccess.

- Line 59-64 : Another block to deny access to backup/logs and other protected file extensions.

About whitespace changes, I use https://github.com/raynigon/vscode-nginx-formatter and it seems the file was fully formatted when I have opened the PR.

Zeokat commented:
`$args` and `$query_string` are [exactly the same](http://nginx.org/en/docs/http/ngx_http_core_module.html#var_query_string), don't need to change it.

The part to deny access to hidden files well... if you are developer you should take care of this. If flarum is focused to the user they shouldn't need this. Also, `.htaccess` files aren't used by Nginx so can't see why block them. In my opinion an overloaded version of nginx configuration doesn't help at all.

Keep things simple is the way to go.

VirtuBox commented:
Hello @Zeokat ,

> `$args` and `$query_string` are [exactly the same](http://nginx.org/en/docs/http/ngx_http_core_module.html#var_query_string), don't need to change it.

There is "**no need**" to change it, but I think it's a good idea to use the same configuration than other popular php applications : 
```nginx
location / {
  try_files $uri $uri/ /index.php?$args;
}
```


> The part to deny access to hidden files well... if you are developer you should take care of this. If flarum is focused to the user they shouldn't need this. Also, `.htaccess` files aren't used by Nginx so can't see why block them. In my opinion an overloaded version of nginx configuration doesn't help at all.
> 

Flarum installation with Nginx is pretty easy, so anybody without strong knowledge in Nginx configuration should be able to use this configuration example by including it a minimal vhost. And that's why (in my opinion) it's better to prevent unwanted access to hidden files than to provide support to users later.

franzliedke commented:
1. Please revert all whitespace and formatting changes.

2. Regarding `expires` - this directives adds the `Cache-Control`, but also the `Expires` header. Do we want this? I am leaning to no.

VirtuBox commented:
> 1. Please revert all whitespace and formatting changes.

Ok, I will revert my commit without the formatting changes 

> 2. Regarding `expires` - this directives adds the `Cache-Control`, but also the `Expires` header. Do we want this? I am leaning to no.

The `Expires`  header do not matter. 
According to Mozilla developers documentation : 
> If there is a Cache-Control header with the max-age or s-maxage directive in the response, the Expires header is ignored.
source : https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires

franzliedke commented on .nginx.conf:
```diff
@@ -1,6 +1,29 @@
+#    server_name forum.domain.tld;
+#    root /var/www/forum.domain.tld/public;
+#    include /var/www/forum.domain.tld/.nginx.conf;
+# }
```
This file is meant to be *included*  in your production config, not used as inspiration, so I am unsure whether this kind of example is useful here - it would probably help more in the documentation. What do others think?

daedaesplayground1991 replied:
Yes definitely but everything is of some help no matter how big or small.

luceos replied:
I agree with @franzliedke ; referencing an include to this file will break environments. Let's not do that. The current version of this file should probably be part of docs not the repo. In addition I'm very unsure why an upstream is being defined globally with a hardcoded php version.

bap14 commented on .nginx.conf:
```diff
@@ -1,6 +1,29 @@
 # Pass requests that don't refer directly to files in the filesystem to index.php
 location / {
-  try_files $uri $uri/ /index.php?$query_string;
+  try_files $uri $uri/ /index.php?$args;
```
Is it not better to use `$is_args` instead of forcing the `?` to be a part of the forwarded URL?

```suggestion
  try_files $uri $uri/ /index.php$is_args$args;
```

VirtuBox replied:
You are totally right, I opened the PR before updating my own nginx configuration for WordPress 5.x, but it's definitively better to use `$is_args$args` insteand of  `?$args`

tobscure commented:
Thanks for your work @VirtuBox. Your changes noticeably increase the *scope* of the included nginx config, in that they include some server config "best practices", which raises an interesting point for discussion:

What is the purpose of this included nginx config file?

In my view it should be to provide the bare minimum config required to make Flarum run with acceptable performance, within an already-functioning Nginx PHP server block. We should not endeavour to provide any further opinionated config or best practices.

bap14 commented:
@tobscure I agree. As an example of an approach I find that works, is Magento 2. They have a sample include file they provide. It does include a commented `server {}` block in the header comments; however, that's only for documentation purposes. The meat of the file uses a variable set in the `server {}` block by the admin to define the "root" of the installation. From there the sample file alters the document root to be the public folder and sets up the "best" routing, and cache / settings, for a base installation. (see https://github.com/magento/magento2/blob/2.3-develop/nginx.conf.sample ).

It's not intrusive, but if the user wants to get up and running quickly, their `server{}` block can be as simple as the following:

    server {
        listen 80;
        server_name myforum.tld;
        
        set $FLARUM_ROOT /var/www/myforum;
        include /var/www/myforum/.nginx.conf;
    }

With the above, they'd be up and running on a Nginx server with what should provide them adequate performance with minimal effort.

VirtuBox commented:
@tobscure if flarum was only a CMS, the minimim nginx config required will be enough.
But flarum is a forum software, with much more sensitive data than a simple CMS.
That's why I added few nginx directives to harden security and avoid possible security issues.

Like @bap14, I think magento 2.X nginx.config.sample is an excellent nginx configuration example, and that's the main inspiration of this PR.

tankerkiller125 commented:
We are closing this due to the lack of recent updates and also too wide of a scope. This might be a better thing to create a tutorial or something of that nature on https://discuss.flarum.org
